### PR TITLE
Backport some compiler performance improvements [ci: last-only]

### PR DESCRIPTION
**Pull request by retronym**
Backports:

  - #7109
  - #7107
  - #7105
  - #7103
  - #7102
  - #7098
  - #7097
  - #7096
  - #6990
  - #6967
  - #6607
  - #6664
  - #6631
  - #6622
  - #6618
  - #6607
  - #6605
  - #6604
  - #6599
  - #6597
  - #6524
  - #6297
  - #6292
  - #6291
  - #6234
  - #6214
  - #6209
  - #7128
  - #7149 
  - #7143

### CODE REVIEW

**Comment by retronym** [comment]
Benchmarks [scheduled](https://scala-ci.typesafe.com/view/scala-bench/job/compiler-benchmark/1822/), results to appear [here](https://scala-ci.typesafe.com/grafana/dashboard/db/pr-validation?orgId=1&from=1501699042299&to=1535585531323&var-scalaSha=e164092%7C4c9780fb7a%7C9a2219405b%7Cf57113ebe6%7Cf7fcf3f908%7C4779c4298c%7C3f651c3916%7C69f32c025d%7C90c94cc2ea%7C9a8328e468%7Cf552875c17&var-source=All&var-bench=HotScalacBenchmark.compile&var-host=scalabench%40scalabench%40)

**Comment by retronym** [comment]
Benchmarks are showing ~10% improvement.

The next candidates for backporting:

 - #7099
 - #7099 
 - #7140 
 - #7142

**Comment by rorygraves** [comment]
Impressive!

**Comment by mkeskells** [comment]
📣

**Comment by milessabin** [comment]
@retronym a word of warning on https://github.com/scala/scala/pull/7147/commits/d0265db8d0ae2c7fa26cfa704570a121201f54cc ... this broke a few things in the 2.13.x community build as discussed [here](https://github.com/scala/scala/pull/6612#issuecomment-402504660). Clearly a bug is being relied on in those cases, nevertheless, including this change will break code which currently compiles on 2.12.6.

If you want to get 95% of the positive effect of that change, then https://github.com/scala/scala/pull/7067 will do it ... if you recall, early pruning eliminated most of the unnecessary thunks.

**Comment by retronym** [comment]
Thanks, @milessabin.

I've whittled the Specs2 thing down to:

```scala
import org.specs2.matcher.{Matcher, ValueCheck, OptionLikeCheckedMatcher, OptionLikeMatcher}


trait Disj[A, B]
class Repro {
  def returnValue[T](check: ValueCheck[T]): Any = null
  implicit def matcherIsValueCheck[T](m: Matcher[T]): ValueCheck[T] = ???

  def overloaded[T](t: ValueCheck[T]): LeftDisjunctionCheckedMatcher[T] = null
  def overloaded[T]: OptionLikeMatcher[({type l[a]= Disj[a, _]})#l, T, T] = null
  trait LeftDisjunctionCheckedMatcher[T] extends OptionLikeCheckedMatcher[({type l[a]=Disj[a, _]})#l, T, T]

  returnValue(overloaded(null: Matcher[AnyRef]))
}
```

Which compiles with 2.12.6, fails with this PR, and works again with this PR with `-Xlog-implicits` which disables the error message optimization.

The `-Xlog-implicits` output is:


```
/Users/jz/code/specs2/scalaz/shared/src/main/scala/org/specs2/matcher/TaskMatchers.scala:13: matcherIsValueCheck is not a valid implicit value for Repro.this.LeftDisjunctionCheckedMatcher[=?T] => org.specs2.matcher.ValueCheck[?] because:
no type parameters for method matcherIsValueCheck: (m: org.specs2.matcher.Matcher[T])org.specs2.matcher.ValueCheck[T] exist so that it can be applied to arguments (Repro.this.LeftDisjunctionCheckedMatcher[?T])
 --- because ---
no unique instantiation of type variable T could be found
  returnValue(overloaded(null: Matcher[AnyRef]))
             ^
/Users/jz/code/specs2/scalaz/shared/src/main/scala/org/specs2/matcher/TaskMatchers.scala:13: matcherIsValueCheck is not a valid implicit value for Repro.this.LeftDisjunctionCheckedMatcher[=?T] => org.specs2.matcher.ValueCheck[?] because:
no type parameters for method matcherIsValueCheck: (m: org.specs2.matcher.Matcher[T])org.specs2.matcher.ValueCheck[T] exist so that it can be applied to arguments (Repro.this.LeftDisjunctionCheckedMatcher[?T])
 --- because ---
no unique instantiation of type variable T could be found
  returnValue(overloaded(null: Matcher[AnyRef]))
             ^
```

I think the implicit search is during static overload resolution.

I'll keep pulling the thread to see what is at fault here.

**Comment by retronym** [comment]
@milessabin Your assessment was spot on, the use of the `=:=` type operations in `explainVariance` was registering a constraint on a type variable in the `pt` of an implicit search.

We can make the test above fail by applying this patch to 2.12.6.

```diff
$ git diff head
diff --git a/src/compiler/scala/tools/nsc/typechecker/Implicits.scala b/src/compiler/scala/tools/nsc/typechecker/Implicits.scala
index de801a3a91..9536ec658c 100644
--- a/src/compiler/scala/tools/nsc/typechecker/Implicits.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/Implicits.scala
@@ -969,7 +969,10 @@ trait Implicits {
               }
             )

+          val mark = undoLog.log
           val typedFirstPending = typedImplicit(firstPending, ptChecked = true, isLocalToCallsite)
+          if (typedFirstPending.isFailure && settings.isScala213)
+            undoLog.undoTo(mark)

           // Pass the errors to `DivergentImplicitRecovery` so that it can note
           // the first `DivergentImplicitTypeError` that is being propagated
```

**Comment by retronym** [comment]
FTR, here's my failed attempt to fully isolate a test case:

```
trait Diffable[T]
object Diffable {
  implicit def universal[T]: Diffable[T] = null
}
class Matcher[-T]
trait ValueCheck[T]
trait BeEqualTypedValueCheck[T] extends ValueCheck[T]
object ValueCheck {
  implicit def typedValueCheck[T](expected: T)(implicit evidence$1: Diffable[T]): BeEqualTypedValueCheck[T] = null
}
trait OptionLikeCheckedMatcher[F[_], T, U] extends Matcher[F[T]]
trait OptionLikeMatcher[F[_], T, U] extends Matcher[F[T]]

trait Disj[A, B]
class Repro {
  def returnValue[T](check: ValueCheck[T]): Any = null
  implicit def matcherIsValueCheck[T](m: Matcher[T]): ValueCheck[T] = ???

  def overloaded[T](t: ValueCheck[T]): LeftDisjunctionCheckedMatcher[T] = null
  def overloaded[T]: OptionLikeMatcher[({type l[a]= Disj[a, _]})#l, T, T] = null
  trait LeftDisjunctionCheckedMatcher[T] extends OptionLikeCheckedMatcher[({type l[a]=Disj[a, _]})#l, T, T]

  returnValue(overloaded(null: Matcher[AnyRef]))
}

```

**Comment by retronym** [comment]
Added a few more, which will add to the [extended benchmark plot](https://scala-ci.typesafe.com/grafana/dashboard/db/pr-validation?orgId=1&from=1501699042299&to=1535585531323&var-scalaSha=5d34c2dfab%7C0e901ba834%7C87bcc75005%7C6be348ac8d%7C6ea993a090%7C03feb56902%7Ce164092a73%7Ce46c4ba144%7Ca857be2e12%7Cb248c1d654%7C2cd6133323%7C346b7ea6a8%7Cad9820d4ff%7C492201babf%7C1a09b1574d%7C799ab6c071%7Cb18f1ac81c%7C00673bbf8f%7C4a55ac81b1%7Cf3c70d9376%7Ce59db3921e%7C81d6b754e5%7Cc5eb150b18%7C1e8480cacf%7C0e579bc13e%7Cf6c5093547%7C798a04e84c%7C4ad2edc4a4%7Cec4846288e%7Cf0c2db828c%7C3ac6c1e422%7Cbe571b66a2%7Cc9f04d0237%7C940e3c2567%7C15f13f1c9e%7Ccc939a52d0%7C1561e1399a%7Ca4fee32dd2%7C6c0b70a81e%7C39a567b8de%7C8398ca5c23%7C833cf7ef13%7C2062b3e63c%7C17fb78c1bf%7Caac607b81f%7Cbd76826ab8%7Caae6a977aa%7Cc1b68e4f4b%7C11ef06f030%7C9e8a421586%7Cc15a386e33%7C32f1944084%7Cd508e7df95%7C9136c5011b%7C90c5264716%7Ce2a0bd9925%7Cc617272912%7Cc3f53687d1%7Cdb73686aaf%7C8a7e43922f&var-source=All&var-bench=HotScalacBenchmark.compile&var-host=scalabench%40scalabench%40)

**Comment by retronym** [comment]
@adriaanm The main thing to review here is the `undoLog` bug fix to `Implicits`, which was motivated by the investigation of the way that the side effects leak was causing specs2 compilation to change course when error message creation was surpressed.

I started out predicating this on `-Xsource:2.13`, but added a commit to turn the fix on unconditionally after we chatted about this and thought it might be such a corner case that we'd be willing to sacrifice bug-compatibility.

**Comment by retronym** [comment]
I wasn't able to identify subset of commits that would give most of the performance benefit, performance ratchets up fairly steadily.

```
c15a386e33fd7ba1b7c1b8635cf9d09d3c6c723f,    838.13,Merge pull request #6697 from lrytz/code-mono
9e8a4215863db97eca7f3f18ae16a88931793934,    828.10,Merge pull request #6655 from hrhino/t10886
11ef06f03048cef40ec3d3210c77a61058fd0a01,    831.65,Merge pull request #6672 from hrhino/t10880
c1b68e4f4bf17be5aaaa2f008d271f26983096ca,    823.90,Merge pull request #6470 from DavidGregory084/bug#10808
aae6a977aadc67b53e75972190327570185fbe03,    835.85,Merge pull request #6439 from som-snytt/issue/10784
bd76826ab8786ce6fd344b4d7eace972bf85cc3c,    828.80,Merge pull request #6712 from exoego/scaladoc-placeholder-color
aac607b81ff5ad9f4402ef2121782cf97be9d4a3,    827.86,Merge pull request #6596 from exoego/older-version-notice
17fb78c1bfa34777cfdaeca0ec65c485a84f96c1,      0.00,Optimize IndexedSeqOptimized.toList
2062b3e63c70d6fad0c87a3c8d317525577441b2,      0.00,Use AnyRefMap in hot parts of the compiler.
833cf7ef13bf65877c3730c41f3fea63468ff863,    827.50,Optimize nested class collection
8398ca5c23a79ac7175e2e0382a1e3e74f695546,    831.60,Optimize generic sig parser
39a567b8defa7eb3f7593774b0dbdd68023825fd,    830.61,Avoid nonEmpty in hot paths
6c0b70a81e79ba5d535aaaa9b1ecfdd714a6ede0,    808.66,Merge branch 'perf-backport/6214-optimizations-in-the-specialization-phase' into perf-backport/omnibus
a4fee32dd2dbf0cbc9b26f170a2c1d160c7f04bc,    802.64,Merge branch 'perf-backport/6234-reduce-overhead-of-ystatistics' into perf-backport/omnibus
1561e1399a384c7067289b73b8d5ebeced588ea5,    808.37,Merge branch 'perf-backport/6291-avoid-allocation-of-classtags-in-hot-code' into perf-backport/omnibus
cc939a52d0ef308e210540867fdbab7a74c95441,    811.80,Merge branch 'perf-backport/6292-array-instantiation' into perf-backport/omnibus
15f13f1c9e887773d91d1884f4ab5891362d67e9,    808.46,Merge branch 'perf-backport/6297-imports-context-lookup' into perf-backport/omnibus
940e3c256721386585dfb96b1f29734b7f2851ab,    813.61,Merge branch 'perf-backport/6524-implicit-prune-bounds' into perf-backport/omnibus
c9f04d023774cc2765a8e16f4593a783de261f80,    804.88,Merge branch 'perf-backport/6597-front-end' into perf-backport/omnibus
be571b66a2b637dea3be239cff2189787abe3e8c,    798.59,Merge branch 'perf-backport/6599-allocation-patmat' into perf-backport/omnibus
3ac6c1e42212566985838104bdc6ff0d459fee57,    790.17,Merge branch 'perf-backport/6604-classtag-unapply-newarray' into perf-backport/omnibus
f0c2db828c202c1c3ef52d64b227837fdcd346c5,    796.63,Merge branch 'perf-backport/6605-methodtype-istrivial' into perf-backport/omnibus
ec4846288e1ec2f750220509c3fc98a8e39ab59d,    791.20,Merge branch 'perf-backport/6607-needles-lub-match' into perf-backport/omnibus
4ad2edc4a42077852a4f13eecfd864ee202a940a,    798.15,Merge branch 'perf-backport/6612-faster-implicit-search-lazy-msg' into perf-backport/omnibus
798a04e84cd1d2bbb2e61e8581cb6e66e248080a,    789.96,Merge branch 'perf-backport/6615-efficiency-scope-lookup' into perf-backport/omnibus
f6c5093547f3d1e021499cf5fb90293045519d49,    792.84,Merge branch 'perf-backport/6616-allocation-typing-transformer' into perf-backport/omnibus
0e579bc13e35ed74bb353199c90435b807792fcd,    783.00,Merge branch 'perf-backport/6617-dead-code' into perf-backport/omnibus
1e8480cacf109e99c8824b67b3c94503ac95fbce,    777.84,Merge branch 'perf-backport/6618-reduce-bulk-typer-inferencer' into perf-backport/omnibus
c5eb150b186c3c1fc4de8c19c230538926cc0571,    770.31,Merge branch 'perf-backport/6622-command-line-parser' into perf-backport/omnibus
81d6b754e5685a5bdb405f6d658ec9edf182412f,    781.56,Merge branch 'perf-backport/6631-footprint-context' into perf-backport/omnibus
e59db3921e19a5c5d4c9326964279e408dfaf8d0,    778.99,Merge branch 'perf-backport/6664-list-apply-rundefinitions' into perf-backport/omnibus
f3c70d9376290b332de7b4f8225ce032e3792ebc,    775.84,[nomerge] Remove test that doesn't apply on 2.12.x
4a55ac81b1a1ca5125fac52e7434d741fa96c787,    767.77,Merge branch 'perf-backport/6607-needles-lub-match' into perf-backport/omnibus
00673bbf8f9e01d3a051890366100ffe7e79b25e,    776.42,Merge branch 'perf-backport/6967-needless-storage-scope' into perf-backport/omnibus
b18f1ac81c58fb8aa1af8c5531ae3885ba866af7,    767.58,Merge branch 'perf-backport/6990-microopts-definitions' into perf-backport/omnibus
799ab6c0710db7dfeee38e058ff03b241a91905f,    762.45,Merge branch 'perf-backport/7096-optimize-method-type-result' into perf-backport/omnibus
1a09b1574d28684d113f7e90cd6ff1f621864297,    758.73,Merge branch 'perf-backport/7097-case-field-accessors' into perf-backport/omnibus
492201babf0cd462074852b9b52a946499344baa,    761.76,Merge branch 'perf-backport/7098-constant-equals-hash' into perf-backport/omnibus
ad9820d4ff682b0ed671116e2aa63e4a85fb6168,    766.90,Merge branch 'perf-backport/7102-erased-types' into perf-backport/omnibus
346b7ea6a85c309570a37313f392f4f1c62747ab,    750.97,Merge branch 'perf-backport/7103-shadower-outer' into perf-backport/omnibus
2cd613332308667e41637a5df559f2502781c6ee,    754.60,Merge branch 'perf-backport/seq-unapply' into perf-backport/omnibus
b248c1d654746ec0a86ae94a7166b00713aaf2ed,    759.32,Merge branch 'perf-backport/7105-remove-indirection-typer' into perf-backport/omnibus
a857be2e121607ae9609c8c6f5578f709d98e498,    763.12,Merge branch 'perf-backport/7107-imports-implicits' into perf-backport/omnibus
e46c4ba14448d841534eacd71de8369bb507065d,    750.16,Merge branch 'perf-backport/7109-pocketful-microopts' into perf-backport/omnibus
e164092a73835b170a6a706f7180008fb1793eb9,    755.62,Address binary incompatibilities
03feb5690248456c2a2db924a4cc32ee27e6f149,    753.98,Merge remote-tracking branch 'origin/2.12.x' into perf-backport/omnibus-first-pr
6ea993a090a34e1c5c53ea38b2c416ae4d1c5d36,    758.05,[mergeforward] Avoid leaking constraints between typechecking implicit candidates
6be348ac8d8ea3ff2fff7ee6d47d6b09de5f0783,      0.00,Merge branch 'perf-backport/7128-temp-string-classpath' into perf-backport/omnibus-first-pr
87bcc750059e0c570dd3ef6166299880fa15f7fd,      0.00,Merge branch 'perf-backport/7149-cold-stats' into perf-backport/omnibus-first-pr
0e901ba834c8c1b8d396c4f04e07b6f7ce576b3b,      0.00,Merge branch 'perf-backport/7143-type-ref-equals' into perf-backport/omnibus-first-pr
5d34c2dfab4ac086079cc5b312187a81a031b61f,    740.20,Address binary incompatibility
0a8e00cb1872e1c032c5f57a447743d27790ba2e,      0.00,[mergeforward] Make nested implicit type error suppression unconditional
da58d8e4f13c75f9ac757ef6f78d40b9cd885c15,      0.00,Merge remote-tracking branch 'origin/2.12.x' into perf-backport/omnibus-first-pr
```

**Comment by retronym** [comment]
This will need a careful merge forward to 2.13.x, I'll volunteer for that job.

**Comment by retronym** [comment]
Running commnunity builds:

 - [2.12.7-bin-5d34c2d-SNAPSHOT](https://scala-ci.typesafe.com/view/scala-2.12.x/job/scala-2.12.x-integrate-community-build/3528)
 - [2.12.7-bin-da58d8e-SNAPSHOT](https://scala-ci.typesafe.com/view/scala-2.12.x/job/scala-2.12.x-integrate-community-build/3527)

**Comment by adriaanm** [comment]
Manually kicked off a build of that last sha, since the list of commits is too long for scabot (haven't implemented paging). https://scala-ci.typesafe.com/view/scala-2.12.x/job/scala-2.12.x-validate-main/1028/console

**Comment by retronym** [comment]
This PR breaks `scalameta` which has a [copy/paste version of vast swathes of Typer](https://github.com/scalameta/scalameta/find/master#L12) that no longer compiles.

@olafurpg We'd need another PR like https://github.com/scalameta/scalameta/pull/1557 to prepare for 2.12.7. Is that something you're able to help out with?

This coupling is really, really undesirable.

File: src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala [hunk 1]
```diff
@@ -201,8 +201,8 @@ abstract class SpecializeTypes extends InfoTransform with TypingTransformers {
     override def run(): Unit = {
       super.run()
       exitingSpecialize {
-        FunctionClass.seq.map(_.info)
-        TupleClass.seq.map(_.info)
+        FunctionClass.seq.take(MaxFunctionAritySpecialized + 1).foreach(_.info)
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala:204]: puzzling :-)

**Review by adriaanm — APPROVED** [review]

File: src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala (hunk 1 above)
Comment (adriaanm) [reply src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala:204]: I understand the reason for MaxFunctionAritySpecialized thanks to the test, but why +1?

**Comment by adriaanm** [comment]
Thanks for dealing with the macro-paradise/semanticdb fallout. Let's get this merged.

File: src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala (hunk 1 above)
Comment (som-snytt) [reply src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala]: There is a Scalafix rule now that calculates the probability of an off-by-one error and injects an adjusting `+ 1` bzw. `- 1`.

**Pull request merged**